src/aspire/abinitio/commonline_c2.py

```diff
@@ -0,0 +1,382 @@
+        pf = self.pf.copy()
+
+        # clmatrix contains the index in image i of the common line with image j
+        # for the two sets of mutual common lines.
+        clmatrix = -np.ones((2, n_img, n_img), dtype=int)
+
+        # 1D shift between common-lines.
+        shifts_1d = np.zeros((2, n_img, n_img))
+
+        # Prepare the shift phases and generate filter for common-line detection.
+        r_max = pf.shape[2]
+        shifts, shift_phases, h = self._generate_shift_phase_and_filter(
+            r_max, self.max_shift, self.shift_step
+        )
+        n_shifts = len(shifts)
+
+        # Apply bandpass filter, normalize each ray of each image.
+        # Note that only use half of each ray.
+        pf = self._apply_filter_and_norm("ijk, k -> ijk", pf, r_max, h)
+
+        # Pre-compute conjugated and shifted pf's.
+        pf_shifted_flipped = np.conj(pf)[:, None] * shift_phases[:, None]
+        pf_shifted_flipped = pf_shifted_flipped.reshape(
+            (self.n_img, n_shifts * (self.n_theta // 2), r_max)
+        )
+
+        # Search for common lines between [i, j] pairs of images.
+        for i in range(n_img - 1):
+            p1 = pf[i]
+            for j in range(i + 1, n_img):
+                p2 = pf_shifted_flipped[j]
+                corr = self._compute_correlations(p1, p2)
+                corr = corr.reshape(self.n_theta, n_shifts, self.n_theta // 2)
+
+                # Find first set of common-lines between the pair of images.
+                first_cl1, first_shift, first_cl2 = np.unravel_index(
+                    corr.argmax(), corr.shape
+                )
+                clmatrix[0, i, j] = first_cl1
+                clmatrix[0, j, i] = first_cl2
+                shifts_1d[0, i, j] = shifts[first_shift]
+
+                # Mask corr around first set of common-lines to search for
+                # second set of mutual common-lines.
+                mask_dist = self.min_dist_cls * 2 * self.n_theta // 360
+                corr_masked = corr * self._square_mask(
+                    corr, first_cl1, first_cl2, mask_dist
+                )
+
+                # Find second set of mutual common-lines.
+                second_cl1, second_shift, second_cl2 = np.unravel_index(
+                    corr_masked.argmax(), corr.shape
+                )
+                clmatrix[1, i, j] = second_cl1
+                clmatrix[1, j, i] = second_cl2
+                shifts_1d[1, i, j] = shifts[second_shift]
+
+        self.clmatrix = clmatrix
+        self.shifts_1d = shifts_1d
+
+    @staticmethod
+    def _compute_correlations(a, b):
+        """
+        Compute the correlation between Polar Fourier images a and b.
+
+        :param a: 2D array size n_theta_a x radial_points.
+        :param b: 2D array size n_theta_b x radial_points.
+
+        :return: Correlation array of size n_theta_a x n_theta_b.
+        """
+        part1 = np.real(a) @ np.real(b).T
```

Comment:
I struggled to come up with better names here. These are the two parts of `np.real(a @ b.T)`.

If we were looking at just a single multiplication `(x + yi)(u + vi) = xu - yv +(xv + yu)i`, then `part1` is `xu` and `part2` is `yv`.

Since we are only using half of the polar Fourier rays, this allows us to build the full correlation by taking `part1 - part2` and `part1 + part2`. 

After explaining this I think this method needs a little better documentation. Also, maybe I should just use `np.real(a @ b.T)` and `np.real(a @ np.conj(b).T)` to compute the correlation in the positive and negative r directions.

Reply:
> Since we are only using half of the polar Fourier rays, this allows us to build the full correlation by taking `part1 - part2` and `part1 + part2`.

Ok. As long as it's documented what's happening, that works.



> After explaining this I think this method needs a little better documentation. Also, maybe I should just use `np.real(a @ b.T)` and `np.real(a @ np.conj(b).T)` to compute the correlation in the positive and negative r directions.

Might be easier to follow if it doesn't slow things down too much.